analyzers/tests/SonarAnalyzer.UnitTest/Helpers/WildcardPatternMatcherTest.cs

```diff
@@ -0,0 +1,114 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.IO;
+
+namespace SonarAnalyzer.UnitTest.Helpers
+{
+    [TestClass]
+    public class WildcardPatternMatcherTest
+    {
+        /// <summary>
+        /// Based on https://github.com/SonarSource/sonar-plugin-api/blob/master/plugin-api/src/test/java/org/sonar/api/utils/WildcardPatternTest.java.
+        /// </summary>
+        [DataTestMethod]
+
+        [DataRow("Foo", "Foo", true)]
+        [DataRow("foo", "FOO", false)]
+        [DataRow("Foo", "Foot", false)]
+        [DataRow("Foo", "Bar", false)]
+
+        [DataRow("org/T?st.java", "org/Test.java", true)]
+        [DataRow("org/T?st.java", "org/Tost.java", true)]
+        [DataRow("org/T?st.java", "org/Teeest.java", false)]
+
+        [DataRow("org/*.java", "org/Foo.java", true)]
+        [DataRow("org/*.java", "org/Bar.java", true)]
+
+        [DataRow("org/**", "org/Foo.java", true)]
+        [DataRow("org/**", "org/foo/bar.jsp", true)]
+
+        [DataRow("org/**/Test.java", "org/Test.java", true)]
+        [DataRow("org/**/Test.java", "org/foo/Test.java", true)]
+        [DataRow("org/**/Test.java", "org/foo/bar/Test.java", true)]
+
+        [DataRow("org/**/*.java", "org/Foo.java", true)]
+        [DataRow("org/**/*.java", "org/foo/Bar.java", true)]
+        [DataRow("org/**/*.java", "org/foo/bar/Baz.java", true)]
+
+        [DataRow("o?/**/*.java", "org/test.java", false)]
+        [DataRow("o?/**/*.java", "o/test.java", false)]
+        [DataRow("o?/**/*.java", "og/test.java", true)]
+        [DataRow("o?/**/*.java", "og/foo/bar/test.java", true)]
+        [DataRow("o?/**/*.java", "og/foo/bar/test.jav", false)]
+
+        [DataRow("org/sonar/**", "org/sonar/commons/Foo", true)]
+        [DataRow("org/sonar/**", "org/sonar/Foo.java", true)]
+
+        [DataRow("xxx/org/sonar/**", "org/sonar/Foo", false)]
+
+        [DataRow("org/sonar/**/**", "org/sonar/commons/Foo", true)]
+        [DataRow("org/sonar/**/**", "org/sonar/commons/sub/Foo.java", true)]
+
+        [DataRow("org/sonar/**/Foo", "org/sonar/commons/sub/Foo", true)]
+        [DataRow("org/sonar/**/Foo", "org/sonar/Foo", true)]
+
+        [DataRow("*/foo/*", "org/foo/Bar", true)]
+        [DataRow("*/foo/*", "foo/Bar", false)]
+        [DataRow("*/foo/*", "foo", false)]
+        [DataRow("*/foo/*", "org/foo/bar/Hello", false)]
+
+        [DataRow("hell?", "hell", false)]
+        [DataRow("hell?", "hello", true)]
+        [DataRow("hell?", "helloworld", false)]
+
+        [DataRow("**/Reader", "java/io/Reader", true)]
+        [DataRow("**/Reader", "org/sonar/channel/CodeReader", false)]
+
+        [DataRow("**", "java/io/Reader", true)]
+
+        [DataRow("**/app/**", "com/app/Utils", true)]
+        [DataRow("**/app/**", "com/application/MyService", false)]
+
```

Comment:
I think we should also document non strictly valid/documented user inputs like these:
```cs
[DataRow("a**/app/**", "aa\\app\\Utils", true)]
[DataRow("a**/app/**", "bb\\app\\Utils", false)]
[DataRow("a**a/app/**", "aa\\app\\Utils", true)]
[DataRow("a**a/app/**", "bb\\app\\Utils", false)]
[DataRow("****/app/**", "aa\\app\\Utils", true)]
[DataRow("****/app/**", "bb\\app\\Utils", true)]
```

Reply:
I would not add this kind of cases. We want to make sure the valid documented scenarios match, and if any other weird not documented pattern matches your case well, and still we should not "document" them as valid scenarios in the UTs.
Why? Our codebase is public and I am afraid if people look at the test cases they will think OK this is a supported good way to add the pattern.

Reply:
I see your point. I ensured that nothing broke on our side since the user input was not validated. Defensive checks/test cases that make sure that everything is working appropriately with wrong input already exists (for ex: poorly formatted xml that should return some default values). Isn't this the same? Maybe to avoid people inferring things just by looking at our codebase we can also add those tests in another method called something like `IsMatch_NonSupportedInput_NotBreak`. WDYT?

Reply:
I added negative test scenarios for "invalid" input, like empty pattern or null.